File: algorithms/java/8_string_to_integer/Atoi.java
class Atoi {
    public int atoi(String str) {
        if (str == null || str.length() == 0) return 0;
        int index = 0;
        while (index < str.length() && str.charAt(index) == ' ') {
            index++;
        }
        if (index == str.length()) return 0;
        int sign = 1;
        if (str.charAt(index) == '-') {
            sign = -1;
            index++;
        } else if (str.charAt(index) == '+') {
            index++;
        }
        int res = 0;
        while (index < str.length() && Character.isDigit(str.charAt(index))) {
            int digit = str.charAt(index) - '0';
            if (res > (Integer.MAX_VALUE - digit) / 10) {
                return sign == -1 ? Integer.MIN_VALUE : Integer.MAX_VALUE;
            }
            res = res * 10 + digit;
            index++;
        }
        return res * sign;
    }

    public static void main(String[] args) {
        Atoi atoi = new Atoi();
        System.out.println(atoi.atoi("130"));
        System.out.println(atoi.atoi("    -128"));
        System.out.println(atoi.atoi("4193 with word"));
        System.out.println(atoi.atoi("word 4193"));
        System.out.println(atoi.atoi("-91283472332")); // underflow return -2147483648
        System.out.println(atoi.atoi("2147483647"));
    }
}

File: algorithms/cpp/8_string_to_integer/Atoi.cpp
class Atoi {
public:
	int myAtoi(string s) {
		if (s.size() == 0) 
            return 0;
        int index = 0;
        while (index < s.size() && s[index] == ' ')
            index++;
        if (index == s.size())
            return 0;
        int sign = 1;
        if (s[index] == '-') {
            sign = -1;
            index++;
        } else if (s[index] == '+') {
            index++;
        }
        int res = 0;
        while (index < s.size() && isdigit(s[index])) {
            int digit = s[index] - '0';
            if (res > (INT_MAX - digit) / 10) {
                return sign == -1 ? INT_MIN : INT_MAX;
            }
            res = res * 10 + digit;
            index++;
        }
        return res * sign;
	}
};
